File: .agents/skills/ros2-pkg-dev/templates/publisher_example.cpp
// ============================================
// C++ ROS2 Publisher 示例 (rclcpp)
// ============================================
// 功能: 定时发布 std_msgs/String 类型消息
// 用法: 参考此代码在现有节点中添加 Publisher
// ============================================

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/string.hpp>

using namespace std::chrono_literals;

class PublisherExample : public rclcpp::Node
{
public:
  PublisherExample()
  : Node("publisher_example"), count_(0)
  {
    // 创建 Publisher
    // 参数: topic名称, QoS队列深度
    publisher_ = this->create_publisher<std_msgs::msg::String>("chatter", 10);

    // 创建定时器，每500ms发布一次消息
    timer_ = this->create_wall_timer(
      500ms, std::bind(&PublisherExample::timer_callback, this));

    RCLCPP_INFO(this->get_logger(), "Publisher 已启动，发布话题: /chatter");
  }

private:
  void timer_callback()
  {
    // 创建消息
    auto message = std_msgs::msg::String();
    message.data = "Hello ROS2! Count: " + std::to_string(count_++);

    // 发布消息
    publisher_->publish(message);

    RCLCPP_INFO(this->get_logger(), "发布: '%s'", message.data.c_str());
  }

  // Publisher 成员变量
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  // 定时器成员变量
  rclcpp::TimerBase::SharedPtr timer_;
  // 计数器
  size_t count_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<PublisherExample>());
  rclcpp::shutdown();
  return 0;
}

File: .agents/skills/ros2-pkg-dev/templates/publisher_example.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
============================================
Python ROS2 Publisher 示例 (rclpy)
============================================
功能: 定时发布 std_msgs/String 类型消息
用法: 参考此代码在现有节点中添加 Publisher
============================================
"""

import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class PublisherExample(Node):
    """Publisher 示例节点"""

    def __init__(self):
        super().__init__('publisher_example')
        self.count = 0

        # 创建 Publisher
        # 参数: 消息类型, topic名称, QoS队列深度
        self.publisher = self.create_publisher(String, 'chatter', 10)

        # 创建定时器，每0.5秒发布一次消息
        self.timer = self.create_timer(0.5, self.timer_callback)

        self.get_logger().info('Publisher 已启动，发布话题: /chatter')

    def timer_callback(self):
        """定时器回调函数"""
        # 创建消息
        msg = String()
        msg.data = f'Hello ROS2! Count: {self.count}'
        self.count += 1

        # 发布消息
        self.publisher.publish(msg)

        self.get_logger().info(f"发布: '{msg.data}'")

    def destroy_node(self):
        """销毁节点"""
        self.get_logger().info('正在销毁 Publisher 节点...')
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = PublisherExample()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('收到键盘中断')
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
